File: CivitasJava/src/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 * @author Miguel Muñoz Molina
 * @author Iván Valero Rodríguez
 */
public class TituloPropiedad {
    private String nombre;
    private float alquilerBase ;
    private static final float factorInteresesHipoteca = (float) 1.1 ;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado ;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario ;
    

    public TituloPropiedad(String nombre, float alquilerBase, float factorRevalorizacion, 
            float hipotecaBase,int precioCompra, int precioEdificar) {
        this.nombre = nombre;
        this.alquilerBase = alquilerBase;
        this.factorRevalorizacion = factorRevalorizacion;
        this.hipotecaBase = hipotecaBase;
        this.hipotecado = false;
        this.numCasas = 0;
        this.numHoteles = 0;
        this.precioCompra = precioCompra;
        this.precioEdificar = precioEdificar;
        this.propietario = null;
    }

    @Override
    public String toString() {
        return "TituloPropiedad{" + "nombre=" + nombre + ", alquilerBase=" + alquilerBase + ", factorRevalorizacion=" + factorRevalorizacion + ", hipotecaBase=" + hipotecaBase + ", hipotecado=" + hipotecado + ", numCasas=" + numCasas + ", numHoteles=" + numHoteles + ", precioCompra=" + precioCompra + ", precioEdificar=" + precioEdificar + ", propietario=" + propietario + '}' + "\n";
    }
    

    
    private float getPrecioAlquiler(){
        float alquiler = 0 ;
        if (!hipotecado && !propietarioEncarcelado())
            alquiler = alquilerBase*(1+(numCasas*(float)0.5)+(numHoteles*(float)2.5)) ;
        return alquiler;
    }
    
    float getImporteCancelarHipoteca(){
        return getImporteHipoteca() * factorInteresesHipoteca ;
    }

    boolean cancelarHipoteca(Jugador jugador){
        boolean hecho = false ;
        if(esEsteElPropietario(jugador) && hipotecado){
            if (jugador.paga(getImporteCancelarHipoteca())){
                hipotecado = false ;
                hecho = true;
            }
        }
        return hecho ;
    }
    
    boolean hipotecar(Jugador jugador){
        boolean hecho = false ;
        if(!hipotecado && esEsteElPropietario(jugador)){
            jugador.recibe(getImporteHipoteca());
            hipotecado = true ;
            hecho = true ;
        }
        return hecho ;
    }
    
    void tramitarAlquiler(Jugador jugador){
        if(tienePropietario() && !esEsteElPropietario(jugador)){
            jugador.pagaAlquiler(getPrecioAlquiler());
            propietario.recibe(getPrecioAlquiler()) ;
        }
    }
    
    private boolean propietarioEncarcelado(){
        if (tienePropietario())
            return propietario.isEncarcelado() ;
        else
            return false ;
    }
    
    int cantidadCasasHoteles(){
        return numCasas + numHoteles ;
    }
    
    boolean derruirCasas(int n, Jugador jugador){
        boolean hecho = false ;
        if(esEsteElPropietario(jugador) && numCasas >= n){
            numCasas -= n ;
            hecho = true ;
        }
        return hecho ;
    }
    
    private float getPrecioVenta(){
        float precioVenta =precioCompra+(numCasas+5*numHoteles)*precioEdificar*factorRevalorizacion ;
        return precioVenta ;
    }
    
    boolean construirCasa(Jugador jugador){
        boolean construida = false ;
        if (esEsteElPropietario(jugador)){
            if(jugador.paga(precioEdificar)){
                numCasas++;
                construida = true ;
            }
        }
        return construida ;
    }
    
    boolean construirHotel(Jugador jugador){
        boolean construido = false ;
        if (esEsteElPropietario(jugador)){
            jugador.paga(3*getPrecioEdificar());
            numHoteles++;
            construido = true ;
            
        }
        return construido ;
    }
    
    boolean Comprar(Jugador jugador){
        boolean comprado = false ;
        if (!tienePropietario()){
            propietario = jugador ;
            comprado = true ;
            jugador.paga(precioCompra) ;
        }
        return comprado ;
    }
    
    boolean tienePropietario(){
        return propietario != null ;
    }
    
    private boolean esEsteElPropietario(Jugador jugador){
        return propietario == jugador ;
    }

    String getNombre() {
        return nombre;
    }

    int getNumCasas() {
        return numCasas;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    float getPrecioCompra() {
        return precioCompra;
    }

    float getPrecioEdificar() {
        return precioEdificar;
    }

    Jugador getPropietario() {
        return propietario;
    }
    
    public boolean getHipotecada(){
        return hipotecado ;
    }
    
    void actualizarPropietarioPorConversion(Jugador jugador){
        throw new UnsupportedOperationException("No implementado") ;
    }
    
    private float getImporteHipoteca(){
        return (float) (hipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5)));
    }
    
    boolean vender(Jugador jugador){
        jugador.modificarSaldo(getPrecioVenta()) ;
        this.derruirCasas(getNumCasas(), jugador) ;
        this.numHoteles = 0 ;
        this.propietario=null;
        return false ;
    }
    
    
    
    
    
}

File: CivitasRuby/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#Author: Miguel Muñoz Molina
#Author: Iván Valero Rodríguez

class TituloPropiedad
  @@factorInteresesHipoteca = 1.1
  
  attr_reader :nombre, :numCasas, :numHoteles,
    :precioCompra, :precioEdificar, :propietario, :hipotecado
  
  def initialize(nombre,alquilerBase,factorRevalorizacion, 
            hipotecaBase,precioCompra,precioEdificar)
          
    @nombre = nombre
    @alquilerBase = alquilerBase
    @factorRevalorizacion = factorRevalorizacion
    @hipotecaBase = hipotecaBase
    @precioCompra = precioCompra
    @precioEdificar = precioEdificar
    @propietario = nil
    @hipotecado = false
    @numCasas = 0
    @numHoteles = 0
  end
  
  def getImporteCancelarHipoteca
    return getImporteHipoteca() * factorInteresesHipoteca ;
  end
  
  def cancelarHipoteca(jugador)
    hecho = false
    if(esEsteElPropietario(jugador) and @hipotecado)
      if (jugador.paga(getImporteCancelarHipoteca))
        @hipotecado = false
        hecho = true
      end
    end
    return hecho
  end
  
  def hipotecar(jugador)
    hecho = false
    if(!@hipotecado and esEsteElPropietario(jugador))
      jugador.recibe(getImporteHipoteca)
      @hipotecado = true 
      hecho = true 
    end
    return hecho 
  end
  
  def tramitarAlquiler(jugador)
    if(tienePropietario and !esEsteElPropietario(jugador))
      jugador.pagaAlquiler(getPrecioAlquiler)
      @propietario.recibe(getPrecioAlquiler)
    end
  end
  
  def cantidadCasasHoteles
    return @numCasas + @numHoteles
  end
  
  def derruirCasas(n, jugador)
    hecho = false
    if(esEsteElPropietario(jugador) and @numCasas >= n)
      @numCasas = @numCasas - n
      hecho = true
    end
    return hecho
  end
  
  def construirCasa(jugador)
    construida = false
    if (esEsteElPropietario(jugador))
      if(jugador.paga(precioEdificar))
        @numCasas = @numCasas + 1
        construida = true
      end
    end
    return construida
  end

  def construirHotel(jugador)
    construido = false
    if (esEsteElPropietario(jugador))
      if(jugador.paga(3*@precioEdificar))
        @numHoteles = @numHoteles + 1
        construido = true
      end
    end
    return construido
  end

  def Comprar(jugador)
    comprado = false
    if (!tienePropietario)
      if(jugador.paga(@precioCompra))
        comprado = true
      end
    end
    return comprado
  end
  
  def tienePropietario
    return @propietario != nil
  end

  #def actualizarPropietarioPorConversion(jugador)
  
  def vender(jugador)
    jugador.modificarSaldo(getprecioVenta())
    derruirCasas(@numCasas,jugador)
    @numHoteles = 0
    return false
  end

  private
  
  def getPrecioAlquiler
    alquiler = 0
    if (!@hipotecado and !propietarioEncarcelado)
      alquiler = @alquilerBase *(1+(@numCasas* 0.5)+(@numHoteles*2.5))
    end
    return alquiler
  end
  
  def propietarioEncarcelado
    if (tienePropietario)
      return @propietario.encarcelado
    else
      return false
    end
  end
  
  def getPrecioVenta
    precioVenta = @precioCompra+(@numCasas+5*@numHoteles)*@precioEdificar*@factorRevalorizacion
    return precioVenta
  end
  
  def esEsteElPropietario(jugador)
    return @propietario == jugador
  end
  
  def getImporteHipoteca
    return (@hipotecaBase*(1+(@numCasas*0.5)+(@numHoteles*2.5)))
  end
  
  def vender(jugador)
    
    if (esEsteElPropietario(jugador) and !@hipotecado )
      jugador.recibe(@precioVenta)
      
      derruirCasas(@numCasas, jugador)
      @numHoteles = 0
      @propietario = nil
      return true
    end
    return false
  end
  
  def to_s
        return "TituloPropiedad{ \n nombre= #{@nombre} , alquilerBase= #{@alquilerBase}, factorRevalorizacion= #{@factorRevalorizacion} , hipotecaBase= #{@hipotecaBase} , hipotecado= #{@hipotecado} , numCasas= #{@numCasas} , numHoteles= #{@numHoteles} , precioCompra= #{@precioCompra} , precioEdificar= #{@precioEdificar} , propietario= #{@propietario} }"
  end
  
  public :to_s
end
